### ISSUE TITLE
Error while trying to refresh Java classpath in Visual Studio Code on MacOS

### ISSUE THREAD

**Issue by TheHandOfNOD** [camel-tooling/vscode-camelk#1698]
I receive the following error when I try to refresh classpath dependencies in Visual Studio Code 1.84.0

I installed the camelK Plugin in version 0.0.36
jbang has the version: 0.111.0 (when running jbang version)
camel JBang version is 4.1.0 via override but the problem also occurs with 4.2.0

```
Found Kubernetes CLI (kubectl) version 1.26... 
Installing kamel... 
Checking to see if Apache Camel K CLI version 2.1.0 available 
Downloading kamel cli tool from https://github.com/apache/camel-k/releases/download/v2.1.0/camel-k-client-2.1.0-darwin-amd64.tar.gz to /Users/daniel/Library/Application Support/Code/User/globalStorage/redhat.vscode-camelk/camelk/tools/kamel/kamel 
Downloading from: https://github.com/apache/camel-k/releases/download/v2.1.0/camel-k-client-2.1.0-darwin-amd64.tar.gz
done 
Error while trying to refresh Java classpath based on file /Users/daniel/.vscode/extensions/redhat.vscode-camelk-0.0.36/resources/simple-java-file/UsedToWarmupDependenciesDownload.java:
Error: Command failed: jbang camel@apache/camel dependency copy --output-directory="/Users/daniel/Library/Application Support/Code/User/globalStorage/redhat.vscode-camelk/java-dependencies" "/Users/daniel/.vscode/extensions/redhat.vscode-camelk-0.0.36/resources/simple-java-file/UsedToWarmupDependenciesDownload.java"
java.nio.file.NoSuchFileException: .camel-jbang/work/src/main/resources/log4j2.properties
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:218)
	at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:484)
	at java.base/java.nio.file.Files.newOutputStream(Files.java:228)
	at java.base/java.nio.file.Files.copy(Files.java:3160)
	at org.apache.camel.dsl.jbang.core.commands.ExportBaseCommand.safeCopy(ExportBaseCommand.java:741)
	at org.apache.camel.dsl.jbang.core.commands.ExportCamelMain.copySourceFiles(ExportCamelMain.java:316)
	at org.apache.camel.dsl.jbang.core.commands.ExportCamelMain.export(ExportCamelMain.java:102)
	at org.apache.camel.dsl.jbang.core.commands.Export.export(Export.java:127)
	at org.apache.camel.dsl.jbang.core.commands.DependencyList.doExport(DependencyList.java:207)
	at org.apache.camel.dsl.jbang.core.commands.DependencyList.export(DependencyList.java:64)
	at org.apache.camel.dsl.jbang.core.commands.ExportBaseCommand.doCall(ExportBaseCommand.java:225)
	at org.apache.camel.dsl.jbang.core.commands.Export.doCall(Export.java:28)
	at org.apache.camel.dsl.jbang.core.commands.DependencyList.doCall(DependencyList.java:54)
	at org.apache.camel.dsl.jbang.core.commands.CamelCommand.call(CamelCommand.java:71)
	at org.apache.camel.dsl.jbang.core.commands.CamelCommand.call(CamelCommand.java:36)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.apache.camel.dsl.jbang.core.commands.CamelJBangMain.run(CamelJBangMain.java:166)
	at main.CamelJBang.main(CamelJBang.java:36)
```

Problem now is that the classpath is not resolved and I end up with a lots of compile errors.

**Comment by lhein** [comment]
Thank you for reporting the issue. We will check it out.

**Comment by apupier** [comment]
I cannot reproduce the issue on Fedora.
We have automated test in success on Mac.
So let's try to find what is specific:
* What is the content of your workspace? Do you reproduce starting from an empty one, create a route using command "Create a Camel route using Java DSL"?
* Do you have the problem when clicking on "refresh classpath dependencies" only? It means when opening the Java file the first time it is working? (after waiting potentially several minutes for jars to be downloaded/copied)

**Comment by TheHandOfNOD** [comment]
Hi, thx for the fast response.

I use Mac Sonoma 14.0 on an M1 Max. 

Exactly. I had/have an empty Visual Studio code open and when I execute the command you mentioned the .java file is created and a little bit later the RouteBuilder and the org.apache import are marked as compilation error because they cannot be resolved. This stays the same also when I restart my VS Code.

The exception stacktrace is only shown when I manually press "Refresh classpath dependencies". 
VS Code is connected to my local Docker Desktop which enables Kubernetes to run. 

Docker Desktop Version: 4.24.2
Kubernetes version v1.27.2

When I execute `kamel get` in the VS code terminal I also see 1 running example which for me indicates that VS code is correctly connected to my local Kubernetes.

**Comment by apupier** [comment]
Can you try to execute the mentioned command in a terminal outside of VS Code (to see if it can come of a problem of execution context inside the VS Code extension):
```jbang camel@apache/camel dependency copy --output-directory="/Users/daniel/Library/Application Support/Code/User/globalStorage/redhat.vscode-camelk/java-dependencies" "/Users/daniel/.vscode/extensions/redhat.vscode-camelk-0.0.36/resources/simple-java-file/UsedToWarmupDependenciesDownload.java"```

Can you check that the `/Users/daniel/.vscode/extensions/redhat.vscode-camelk-0.0.36/resources/simple-java-file/UsedToWarmupDependenciesDownload.java` file exists?



> I discovered a .camel-jbang folder now in my VSCode which contains a camel-jbang-run.properties file with the following settings in it:

so it sounds like it is using Camel 4.0.0.
Can you try to delete this folder? It it recreated the same? If yes, can you check the global camel config `jbang camel@apache/camel config list`?

**Comment by TheHandOfNOD** [comment]
Hi, thx for the response.

I executed the command in a terminal outside of VSCode

```
jbang camel@apache/camel dependency copy --output-directory="/Users/daniel/Library/Application Support/Code/User/globalStorage/redhat.vscode-camelk/java-dependencies" "/Users/daniel/.vscode/extensions/redhat.vscode-camelk-0.0.36/resources/simple-java-file/UsedToWarmupDependenciesDownload.java"
[jbang] Resolving dependencies...
[jbang]    org.apache.camel:camel-bom:4.2.0@pom
[jbang]    org.apache.camel:camel-jbang-core:4.2.0
[jbang]    org.apache.camel.kamelets:camel-kamelets:4.2.0
[jbang] Dependencies resolved
[jbang] Building jar for CamelJBang.java...
[jbang] There is a new version of jbang available!
[jbang] You have version 0.111.0 and 0.112.4 is the latest.
[jbang] Run 'jbang version --update' to update to the latest version.
```

The file does indeed exist.

I deleted the .camel-jbang folder in my vscode and it does not seem to be recreated (I also restarted VSCode and pressed on the refresh button several times in the Explorer-view inside VSCode.

The output is still the same:

```
Found Kubernetes CLI (kubectl) version 1.26... 
Installing kamel... 
Checking to see if Apache Camel K CLI version 2.1.0 available 
Downloading kamel cli tool from https://github.com/apache/camel-k/releases/download/v2.1.0/camel-k-client-2.1.0-darwin-amd64.tar.gz to /Users/daniel/Library/Application Support/Code/User/globalStorage/redhat.vscode-camelk/camelk/tools/kamel/kamel 
Downloading from: https://github.com/apache/camel-k/releases/download/v2.1.0/camel-k-client-2.1.0-darwin-amd64.tar.gz
done 
Error while trying to refresh Java classpath based on file /Users/daniel/.vscode/extensions/redhat.vscode-camelk-0.0.36/resources/simple-java-file/UsedToWarmupDependenciesDownload.java:
Error: Command failed: jbang camel@apache/camel dependency copy --output-directory="/Users/daniel/Library/Application Support/Code/User/globalStorage/redhat.vscode-camelk/java-dependencies" "/Users/daniel/.vscode/extensions/redhat.vscode-camelk-0.0.36/resources/simple-java-file/UsedToWarmupDependenciesDownload.java"
java.nio.file.NoSuchFileException: .camel-jbang/work/src/main/resources/log4j2.properties
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:218)
	at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:484)
	at java.base/java.nio.file.Files.newOutputStream(Files.java:228)
	at java.base/java.nio.file.Files.copy(Files.java:3160)
	at org.apache.camel.dsl.jbang.core.commands.ExportBaseCommand.safeCopy(ExportBaseCommand.java:741)
	at org.apache.camel.dsl.jbang.core.commands.ExportCamelMain.copySourceFiles(ExportCamelMain.java:316)
	at org.apache.camel.dsl.jbang.core.commands.ExportCamelMain.export(ExportCamelMain.java:102)
	at org.apache.camel.dsl.jbang.core.commands.Export.export(Export.java:127)
	at org.apache.camel.dsl.jbang.core.commands.DependencyList.doExport(DependencyList.java:207)
	at org.apache.camel.dsl.jbang.core.commands.DependencyList.export(DependencyList.java:64)
	at org.apache.camel.dsl.jbang.core.commands.ExportBaseCommand.doCall(ExportBaseCommand.java:225)
	at org.apache.camel.dsl.jbang.core.commands.Export.doCall(Export.java:28)
	at org.apache.camel.dsl.jbang.core.commands.DependencyList.doCall(DependencyList.java:54)
	at org.apache.camel.dsl.jbang.core.commands.CamelCommand.call(CamelCommand.java:71)
	at org.apache.camel.dsl.jbang.core.commands.CamelCommand.call(CamelCommand.java:36)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.apache.camel.dsl.jbang.core.commands.CamelJBangMain.run(CamelJBangMain.java:166)
	at main.CamelJBang.main(CamelJBang.java:36)
```

I executed the command `jbang camel@apache/camel config list` outside of VS code and it returns:
camel-version = 4.1.0

In the terminal inside VSCode it's the same output.

**Comment by djelinek** [comment]
Hello @TheHandOfNOD ,
we were finally able to reproduce locally on my macOS laptop. We have tested `jbang dependency copy` command manually inside VS Code integrated `zsh` and `bash` terminals and also outside VS Code where everything was working fine for both environments.

VS Code Camel K extension is using `execSync` from npm `child_process` which is only different we identified that could be problematic for some reason when no current working directory (`cwd`) is not specified for this process execution (see [L54](https://github.com/camel-tooling/vscode-camelk/blob/main/src/JavaDependenciesManager.ts#L54) of `JavaDependenciesManager` class implementation)

```typescript
execSync(command);
```

on behalf of all of these and previous findings, @apupier has provided way how to fix by specifying the current working directory, which we have successfully tested that after there is no more error spawned when opening Java DSL route.

```typescript
execSync(command, { cwd: (vscode.workspace.workspaceFolders as vscode.WorkspaceFolder[])[0].uri.fsPath });
```
